### ISSUE TITLE
MiniJAXB should not throw NoClassDefFoundError in the absence of the JAXB API on classpath

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#9723]
When users do not use Maven to build jOOQ projects (e.g. as suggested in the tutorial), there is currently a `NoClassDefFoundError: javax/xml/bind/annotation/XmlSchema`.

This particular exception was introduced with https://github.com/jOOQ/jOOQ/issues/8918. However, other similar exceptions could have been caused before, because we used to use ordinary reflection API to read JAXB API annotations before.

`MiniJAXB` should be able to work without the JAXB API on the classpath.

The JAXB API ships with the CDDL license. I'm not sure if that license is compatible with ASL 2.0, so I'm not sure whether we can ship the API with the jOOQ ZIP file for convenience.

----
See also:
- https://stackoverflow.com/q/59690301/521799
- https://www.jooq.org/doc/latest/manual/code-generation/codegen-configuration/#comment-4752785628

**Comment by lukaseder** [comment]
As we are currently using Maven throughout our integration tests, this has gone undetected. There's an integration test project `jOOQ-test-codegen-no-jaxb`, which avoids the JAXB *implementation* dependency to test whether `MiniJAXB` works correctly. But we do not yet have an integration test that avoids the JAXB *API*.

I suspect we cannot use ordinary reflection API to read annotations that are not on the classpath.

**Comment by lukaseder** [comment]
Without the dependency on the classpath, the following code doesn't print anything:

```java
public static void main(String[] args) {
    Class<Settings> s = Settings.class;

    for (var a : s.getAnnotations()) {
        System.out.println(a);
    }
}
```

I.e. the annotations are not visible to the reflection API if their classes cannot be loaded. When explicitly adding the JAXB API to the classpath, we get:

```
@javax.xml.bind.annotation.XmlAccessorType(value=FIELD)
@javax.xml.bind.annotation.XmlType(name="Settings", factoryMethod="", factoryClass=javax.xml.bind.annotation.XmlType$DEFAULT.class, namespace="##default", propOrder={})
```

A workaround would be to duplicate all the JAXB annotations that we need in some `org.jooq.minijaxb` package and extend our XJC code generation to produce both the standard JAXB annotations as well as our own...

**Comment by knutwannheden** [comment]
In the meantime we should maybe document in the manual that for Java 11 the generator requires the JAXB API on the classpath.

**Comment by lukaseder** [comment]
I intentionally left that out because there might be a fix that allows avoiding this entirely.

If not, I wonder if we should burden newbies (who might also be Java newbies) with the task of finding a way to download the JAXB API jar file manually. After all, this is a tutorial and things should be easy.

**Comment by lukaseder** [comment]
This has now been addressed via https://github.com/jOOQ/jOOQ/issues/12286:

- The JAXB API `.jar` and `-sources.jar` is distributed with the various ZIP files
- Including that API is documented in the tutorial: https://www.jooq.org/doc/latest/manual/getting-started/tutorials/jooq-in-7-steps/jooq-in-7-steps-step3 (the tutorial already contains quite a few jar files, so adding one more doesn't hurt)
- Maven/Gradle/jbang users are not affected by this, as the dependency is downloaded automatically

So, given that the only users who are running into this are users of the ZIP file that do not use Maven, Gradle, or jbang, and that for those users, the issue #12286 has now been fixed, I don't think we need to implement any further fixes here.

**Comment by lukaseder** [comment]
Duplicate of #12286

**Comment by louwers** [comment]
Ironically, I am getting this issue on `3.16.2` but not on `3.15.0`.

**Comment by jehrlich** [comment]
You help no one by publishing a tutorial that does not work as written. A warning in the tutorial is needed, at least. It bit me almost enough to make me give up entirely.

**Comment by lukaseder** [comment]
Thanks for your messages folks.

@louwers: Yes there seems to be an incompatibility in that third party contribution introduced by a change in jOOQ 3.16, probably this one: https://github.com/jOOQ/jOOQ/issues/9641. I'm looking into resolving this ASAP, together with the contributor of that plugin, @etiennestuder

@jehrlich Thanks for your patience. There's no warning needed, but fixes. I'm sorry to hear this will have been too late for you, but your feedback here and in the other comment you've raised (https://github.com/jOOQ/jOOQ/issues/12286#issuecomment-1026403699), will definitely help others, in the near future.

**Comment by lukaseder** [comment]
The manual will have to be fixed: https://github.com/jOOQ/jOOQ/issues/12983

**Comment by perlun** [comment]
> Yes there seems to be an incompatibility in that third party contribution introduced by a change in jOOQ 3.16, probably this one: #9641. I'm looking into resolving this ASAP, together with the contributor of that plugin, @etiennestuder

For the record, the following fixes in the plugin makes it compliant with jOOQ 3.16. :tada: 

* https://github.com/etiennestuder/gradle-jooq-plugin/pull/206
* https://github.com/etiennestuder/gradle-jooq-plugin/commit/1c84650c8e530a8155dc205a16d0c3681e5b6828

There is not yet any release of the `gradle-jooq-plugin` with these fixes; I will kindly ask @etiennestuder for a new release. :+1:

**Comment by etiennestuder** [comment]
The very recent [v7.0 of the gradle-jooq-plugin](https://github.com/etiennestuder/gradle-jooq-plugin/releases/tag/v7.0) fixes the situation and is compatible with 3.16.x. People still using jOOQ <=3.15 should continue to use v6.0.1.
